File: core/FindServersThatHandledMostNumberOfRequests/FindServersThatHandledMostNumberOfRequests.py
from typing import List
from collections import Counter
import heapq

def busiestServers(k: int, arrival: List[int], load: List[int]) -> List[int]:
    available = list(range(k))
    busy = []
    requests = [0] * k
    for i, (start, t) in enumerate(zip(arrival, load)):
        while busy and busy[0][0] <= start:
            _, id = heapq.heappop(busy)
            heapq.heappush(available, i + (id - i) % k)
        if available:
            id = heapq.heappop(available) % k
            requests[id] += 1
            heapq.heappush(busy, (start + t, id))
    res = max(requests)
    return [i for i, req in enumerate(requests) if req == res]


if __name__ == '__main__':
    print(busiestServers(k = 3, arrival = [1,2,3,4,5], load = [5,2,3,3,3] ))

File: core/FindServersThatHandledMostNumberOfRequests/FindServersThatHandledMostNumberOfRequests.java
package com.dbc;

import java.util.*;

public class FindServersThatHandledMostNumberOfRequests {
    public List<Integer> busiestServers(int k, int[] arrival, int[] load) {
        int[] requests = new int[k];
        Queue<Integer> available = new PriorityQueue<Integer>((a, b) -> a - b);
        for (int i = 0; i < k; i++) available.offer(i);
        Queue<int[]> busy = new PriorityQueue<int[]>((a, b) -> a[0] - b[0]);
        for (int i = 0; i < arrival.length; i++) {
            while (!busy.isEmpty() && busy.peek()[0] <= arrival[i]) {
                int id = busy.peek()[1];
                busy.poll();
                available.offer(i + ((id - i) % k + k) % k);
            }
            if (available.isEmpty()) continue;
            int server = available.poll() % k;
            requests[server]++;
            busy.offer(new int[]{arrival[i] + load[i], server});
        }
        int max = Arrays.stream(requests).max().getAsInt();
        List<Integer> res = new ArrayList<>();
        for (int i = 0; i < k; i++) {
            if (requests[i] == max) {
                res.add(i);
            }
        }
        return res;
    }
}
